Store L3 temporary files in temporary directory, take 2

mrandri19 opened the pull request.

mrandri19 committed:
Store L3 temporary files in temporary directory

g8a9 commented on l3wrapper/l3wrapper.py:
```diff
@@ -294,6 +300,9 @@ def fit(self, X, y, column_names=None, save_human_readable=False, remove_files=T
         if remove_files:
             _remove_fit_files(token)
 
+        chdir(old_dir)
+        tmp_dir.cleanup()
+
```
I can still see some problems here. If a client chooses to **save_human_readable** format of the rules, the cleaning process of the temporary directory should not happen, otherwise, we lose them.

Maybe something like that?

```suggestion
        if remove_files:
            _remove_fit_files(token)

        chdir(old_dir)
        if remove_files and not save_human_readable:
            tmp_dir.cleanup()
```

g8a9 commented on l3wrapper/l3wrapper.py:
```diff
@@ -232,6 +234,10 @@ def fit(self, X, y, column_names=None, save_human_readable=False, remove_files=T
         self.X_ = X
         self.y_ = y
 
+        tmp_dir = TemporaryDirectory()
```
The point behind the **current_token_** is to help clients to keep track of the results generated by the last call to the **fit()** method. Now, every file that starts with *token_* is generated inside the temporary directory which makes it harder to access.

It might be useful to add the generated token to the name of the tmp_dir itself.
 ```suggestion
        token = secrets.token_hex(4)
        filestem = f"{token}"
        
        tmp_dir = TemporaryDirectory(prefix=token)
        old_dir = getcwd()
        chdir(tmp_dir.name)
```

mrandri19 replied:
Ah, I understand now, yes, this seems like a good solution

g8a9 reviewed (changes requested):
I've left you a couple of code reviews. Let me know how do you see them.

mrandri19 committed:
Update l3wrapper/l3wrapper.py

Co-authored-by: Giuseppe Attanasio <giuseppeattanasio6@gmail.com>

mrandri19 committed:
Update l3wrapper/l3wrapper.py

Co-authored-by: Giuseppe Attanasio <giuseppeattanasio6@gmail.com>

g8a9 commented on l3wrapper/l3wrapper.py:
```diff
@@ -232,6 +234,15 @@ def fit(self, X, y, column_names=None, save_human_readable=False, remove_files=T
+       tmp_dir = TemporaryDirectory(prefix=token)
+       old_dir = getcwd()
+       chdir(tmp_dir.name)
+        old_dir = getcwd()
```
I do not know if I have a problem on my end. Is this a duplicate?

g8a9 reviewed (commented):
I have left one more comment.

**EDIT:** I added one commit to address the issues. It seems right to me.

g8a9 committed:
Fixing spaces and duplicated lines in fit()

g8a9 committed:
Update the logic for the training directory.

Since the directory created for the training files in not "temporary" in some cases, it is better to use the current token only and eventually remove it if is needed.

g8a9 committed:
Add and improve test cases for the new logic of training.
